runtime/src/main/kotlin/konan/worker/Lazy.kt

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright 2010-2018 JetBrains s.r.o.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package konan.worker
+
+internal class FreezeAwareLazyImpl<out T>(initializer: () -> T) : Lazy<T> {
+    private var initializer_: (() -> T)? = initializer
+    private var value_: Any? = UNINITIALIZED
+    private val valueFrozen_ = AtomicReference<Any?>(UNINITIALIZED)
+
+    override val value: T
+        get() {
+            var result: Any? = value_
+            if (isFrozen) {
+                if (result !== UNINITIALIZED) {
+                    assert(result !== INITIALIZING)
+                    @Suppress("UNCHECKED_CAST")
+                    return result as T
+                }
+                // Barrier.
+                do {
+                    result = valueFrozen_.get()
+                } while (result === INITIALIZING)
```

Comment:
In case of recursive initialization we seem to get infinite loop here.

Reply:
Yes, as designed.

Reply:
Hung program is harder to debug than crashed one.

Reply:
I have quite an opposite opinion, if debugger is available. Anyway, what do you suggest?

Reply:
I suggest to think a little about user-friendly handling for this case.

Btw, spin lock doesn't seem to be most suitable mutex implementation for this class. On the other hand, we are still missing proper lock implementations in our stdlib, which would be useful here and in many other cases.